add navigation for search feature

yousef-osama11 opened the pull request.

yousef-osama11 committed:
add navigation for search feature

Joseph-Sameh-0 committed:
Convert searchUi from library to application module

The searchUi module has been converted from an Android library to an Android application.
This change allows the module to be run as a standalone application.

Key changes include:
- Modified `build.gradle.kts` to use the `android.application` plugin and set application-specific configurations like `applicationId`, `targetSdk`, `versionCode`, and `versionName`.
- Updated `AndroidManifest.xml` to include an intent filter for the main activity, making it the launcher activity.
- Added necessary resources like launcher icons, colors, and updated string resources.
- Removed `consumer-rules.pro` as it's not needed for an application module.
- Updated `Destinations.kt` to make `SearchScreen` a data object for serialization.
- Corrected the package name in `ExampleInstrumentedTest.kt`.

Joseph-Sameh-0 committed:
refactor: Relocate SearchApp and add Kotlin serialization plugin

- Moved `SearchApp.kt` from `navigation` package to `searchUi` package.
- Updated `AndroidManifest.xml` to reflect the new path of `SearchApp`.
- Added the Kotlin serialization plugin to `build.gradle.kts`.

mohammed-akkad commented on feature/search/searchUi/src/main/java/com/feature/search/searchUi/navigation/SearchNavHost.kt:
```diff
@@ -0,0 +1,28 @@
+@Composable
+fun SearchNavHost(navController: NavHostController) {
+    NavHost(navController = navController, startDestination = Destinations.SearchScreen) {
+        composable<Destinations.SearchScreen> {
```
You can make file like searhRoute
And content extension function 
Like this
```
fun NavGraphBuilder.searchScreen(){
        composable<Destinations.SearchScreen> { 
             SearchScreen() 
         }
}
```
This approach improves modularity and maintainability.

Joseph-Sameh-0 committed:
refactor: Implement Navigator pattern for Search feature

This commit refactors the navigation within the Search feature to use a Navigator pattern.

Key changes:
- Introduced `Navigator`, `NavigatorImpl`, `NavigationEvent`, `Graph`, and `Destination` interfaces/classes to manage navigation logic.
- Implemented `ObserveAsEvents` Composable to handle navigation events.
- Updated `SearchNavGraph` to use the new Navigator and observe navigation events.
- Removed `SearchNavHost` and `LocalNavController` as they are no longer needed.
- Modified ViewModels to use the injected `Navigator` for navigation actions.
- Updated Screens to call ViewModel methods for navigation.
- Registered `NavigatorImpl` in the Koin dependency injection module.
- Adjusted `Destinations` to implement the new `Graph` and `Destination` interfaces.
- FilterScreen is now a dialog.

Joseph-Sameh-0 committed:
Merge pull request #15 from Paris-Squad-S2/refactor/update-navigation

refactor: Implement Navigator pattern for Search feature

Joseph-Sameh-0 committed:
Merge branch 'develop' into feature/search_navigation

Joseph-Sameh-0 committed:
refactor: Create navigation graph for search feature

This commit introduces a dedicated navigation graph for the search feature.

The `buildSearchNavGraph` function is now defined in `Destinations.kt` and utilized within `SearchNavGraph.kt` to construct the navigation flow. This centralizes the search feature's navigation logic.

Joseph-Sameh-0 committed:
delete UI theme and type files

Removed unnecessary UI theme and type files from the `searchUi` feature.
Also, removed a comment from FindByActorScreen, WorldTourScreen, and SearchScreen, and unused imports from MainActivity.
